File: src/backend/app/schemas/ai.py
from typing import List, Literal, Optional

from pydantic import BaseModel, Field


class QuestionOptions(BaseModel):
    A: str = Field(..., description="Option A")
    B: str = Field(..., description="Option B")
    C: str = Field(..., description="Option C")
    D: str = Field(..., description="Option D")


class AssessmentQuestion(BaseModel):
    id: int = Field(..., description="Question ID or sequence number")
    question: str = Field(..., description="The multiple choice question text")
    options: QuestionOptions = Field(
        ..., description="Dictionary containing options A, B, C, and D"
    )
    correct_answer: Literal["A", "B", "C", "D"] = Field(
        ..., description="The correct option key (A, B, C, or D)"
    )
    explanation: str = Field(
        ..., description="Detailed explanation of the correct answer"
    )


class SkillAssessmentResponse(BaseModel):
    competency: str = Field(..., description="Name of the competency")
    category: str = Field(..., description="Category of the competency")
    department: str = Field(..., description="Target department")
    level: int = Field(..., description="Proficiency level")
    questions: List[AssessmentQuestion] = Field(
        ..., description="List of multiple-choice assessment questions"
    )


class SkillAssessmentQuery(BaseModel):
    competency: str = Field(..., description="Name of the competency")
    description: str = Field(..., description="Description of the category")
    category: str = Field(..., description="Category of the competency")
    department: str = Field(..., description="Target department")
    level: int = Field(..., description="Proficiency level")


class QuizOption(BaseModel):
    id: Literal["a", "b", "c", "d"] = Field(
        ..., description="Option identifier (a, b, c, or d)"
    )
    text: str = Field(..., description="Option text")


class DocumentQuizQuestion(BaseModel):
    text: str = Field(..., description="The multiple choice question text")
    options: List[QuizOption] = Field(
        ..., description="Array of 4 options (a, b, c, d)"
    )
    correctOptionId: Literal["a", "b", "c", "d"] = Field(
        ..., description="The correct option identifier"
    )
    explanation: str = Field(
        ..., description="Explanation for why the correct answer is right"
    )
    difficulty: Literal["easy", "medium", "hard"] = Field(
        ..., description="Difficulty level"
    )
    competencyTag: str = Field(..., description="Associated competency area tag")


class DocumentQuizResponse(BaseModel):
    quizTitle: str = Field(..., description="Title of the generated quiz")
    questions: List[DocumentQuizQuestion] = Field(
        ..., description="List of generated quiz questions"
    )


class ChatMessage(BaseModel):
    role: Literal["user", "assistant", "model", "system"] = Field(
        ..., description="Role of the message sender"
    )
    content: str = Field(
        ..., min_length=1, max_length=10000, description="Message text content"
    )


class AIChatRequest(BaseModel):
    messages: List[ChatMessage] = Field(
        ..., min_length=1, description="Conversation history with the user"
    )
    uploaded_material_text: Optional[str] = Field(
        None, max_length=20000, description="Optional active document context"
    )


class ActionExecuted(BaseModel):
    type: Literal["enroll_course"]
    courseId: str
    courseTitle: Optional[str] = None
    officialId: str


class AIChatResponse(BaseModel):
    reply: str = Field(..., description="Conversational reply text")
    action_executed: Optional[ActionExecuted] = None
    action_status: Optional[
        Literal["success", "already_enrolled", "not_found", "error"]
    ] = None
    action_message: Optional[str] = None

File: src/frontend/src/services/ai.ts
import { apiRequest } from "./api";

export interface QuestionOptions {
  A: string;
  B: string;
  C: string;
  D: string;
}

export interface AssessmentQuestion {
  id: number;
  question: string;
  options: QuestionOptions;
  correct_answer: "A" | "B" | "C" | "D";
  explanation: string;
}

export interface SkillAssessmentResponse {
  competency: string;
  category: string;
  department: string;
  level: number;
  questions: AssessmentQuestion[];
}

export interface SkillAssessmentQuery {
  competency: string;
  description: string;
  category: string;
  department: string;
  level: number;
}

export interface QuizOption {
  id: "a" | "b" | "c" | "d";
  text: string;
}

export interface DocumentQuizQuestion {
  text: string;
  options: QuizOption[];
  correctOptionId: "a" | "b" | "c" | "d";
  explanation: string;
  difficulty: "easy" | "medium" | "hard";
  competencyTag: string;
}

export interface DocumentQuizResponse {
  quizTitle: string;
  questions: DocumentQuizQuestion[];
}

export interface ChatMessage {
  role: "user" | "assistant" | "model" | "system";
  content: string;
}

export interface ActionExecuted {
  type: "enroll_course";
  courseId: string;
  courseTitle?: string;
  officialId: string;
}

export interface AIChatResponse {
  reply: string;
  action_executed?: ActionExecuted | null;
  action_status?: "success" | "already_enrolled" | "not_found" | "error" | null;
  action_message?: string | null;
}

export interface AIChatRequest {
  messages: ChatMessage[];
  uploaded_material_text?: string | null;
}

export const aiService = {
  async getCompetencyQuiz(
    query: SkillAssessmentQuery,
  ): Promise<SkillAssessmentResponse> {
    const params = new URLSearchParams({
      competency: query.competency,
      description: query.description || "",
      category: query.category,
      department: query.department,
      level: query.level.toString(),
    });

    return apiRequest<SkillAssessmentResponse>(
      `/ai/getquiz?${params.toString()}`,
      {
        method: "GET",
      },
    );
  },

  async generateDocumentQuiz(
    file: File,
    numQuestions: number = 5,
    difficulty: string = "mixed",
    competencyTags?: string[],
  ): Promise<DocumentQuizResponse> {
    const formData = new FormData();
    formData.append("file", file);
    formData.append("num_questions", numQuestions.toString());
    formData.append("difficulty", difficulty);
    if (competencyTags && competencyTags.length > 0) {
      formData.append("competency_tags", competencyTags.join(", "));
    }

    return apiRequest<DocumentQuizResponse>("/ai/generate-document-quiz", {
      method: "POST",
      body: formData,
    });
  },

  async chatWithAssistant(request: AIChatRequest): Promise<AIChatResponse> {
    return apiRequest<AIChatResponse>("/ai/chat", {
      method: "POST",
      body: JSON.stringify(request),
    });
  },
};
